File: src/pkg_cpp_py/scripts/pub.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from interfaces.msg import Num

class Pub(Node) :
    def __init__(self) :
        super().__init__("publisher")
        self.publisher_ = self.create_publisher(Num, "topic", 10)
        self.msg = Num()
        self.count_ = 0
        self.timer_ = self.create_timer(1.0, self.callback_timer)
        self.get_logger().info("Start publisher")

    def callback_timer(self) :
        self.msg.num = self.count_
        self.count_ += 1
        self.publisher_.publish(self.msg)

def main(args=None) :
    rclpy.init(args=args)
    node = Pub()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__" :
    main()

File: src/pkg_cpp_py/src/pub.cpp
#include "../include/pkg_cpp_py/pub.hpp"

Pub::Pub() : Node("publisher"), count_(0)
{
    this->publisher_ = this->create_publisher<interfaces::msg::Num>("topic", 10);
    this->timer_ = this->create_wall_timer(1s, std::bind(&Pub::callback_timer, this));
    RCLCPP_INFO(this->get_logger(), "Start publisher");
}

void Pub::callback_timer()
{
    auto message = interfaces::msg::Num();
    message.num = this->count_++;
    this->publisher_->publish(message);
}

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Pub>());
    rclcpp::shutdown();

    return 0;
}
